src/blocks/ibus.rs

```diff
@@ -0,0 +1,185 @@
+            .block_error("ibus", "failed to acquire lock")?)
+            .clone();
+        self.text.set_text(string);
+        Ok(None)
+    }
+
+    // Returns the view of the block, comprised of widgets.
+    fn view(&self) -> Vec<&I3BarWidget> {
+        vec![&self.text]
+    }
+
+    // This function is called on every block for every click.
+    // TODO: Filter events by using the event.name property,
+    // and use to switch between input engines?
+    fn click(&mut self, _: &I3BarEvent) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn parse_msg(ci: &ConnectionItem) -> Option<&str> {
+    let m = if let &ConnectionItem::Signal(ref s) = ci { s } else { return None };
+    if &*m.interface().unwrap() != "org.freedesktop.IBus" { return None };
+    if &*m.member().unwrap() != "GlobalEngineChanged" { return None };
+    let engine = m.get1::<&str>();
+    engine
+}
+
+// Gets the address being used by the currently running ibus daemon.
+// 
+// By default ibus will write the address to `$XDG_CONFIG_HOME/ibus/bus/aaa-bbb-ccc`
+// where aaa = dbus machine id, usually found at /etc/machine-id
+//       bbb = hostname - seems to be "unix" in most cases [see L99 of reference]
+//       ccc = display number from $DISPLAY
+// Refer to: https://github.com/ibus/ibus/blob/7cef5bf572596361bc502e8fa917569676a80372/src/ibusshare.c
+//
+// Example file contents:
+// ```
+// # This file is created by ibus-daemon, please do not modify it
+// IBUS_ADDRESS=unix:abstract=/tmp/dbus-8EeieDfT,guid=7542d73dce451c2461a044e24bc131f4
+// IBUS_DAEMON_PID=11140
+// ```
+fn get_ibus_address() -> Result<String> {
+    // TODO: Check IBUS_ADDRESS variable, as it seems it can be manually set too.
+
+    let config_dir = env::var("XDG_CONFIG_HOME")
+        .block_error("ibus", "Problem getting $XDG_CONFIG_HOME")?;
+
+    // TODO: Check /var/lib/dbus/machine-id if /etc/machine-id fails
+    let mut f = File::open("/etc/machine-id")
+        .block_error("ibus", "Could not open /etc/machine-id")?;
+    let mut machine_id = String::new();
+    f.read_to_string(&mut machine_id)
+        .block_error("ibus", "Something went wrong reading /etc/machine-id")?;
+    let machine_id = machine_id.trim();
+
+    // On sway, $DISPLAY is only set by programs requiring xwayland, such as ibus (GTK2).
+    // ibus-daemon can be autostarted by sway (via an entry in config file), however since
+    // the bar is executed first, $DISPLAY will not yet be set at the time this code runs.
+    // Hence on sway you will need to reload the bar once after login to get the block to work.
+    let display = env::var("DISPLAY")
+        .block_error("ibus", "$DISPLAY not set. Try restarting block if on sway")?;
+    let re = Regex::new(r"^:(\d{1})$").unwrap();
+    let cap = re.captures(&display).unwrap();
+    let display = &cap[1].to_string();
+
+    let hostname = String::from("unix");
+
+    let ibus_socket_path = format!("{}/ibus/bus/{}-{}-{}", config_dir, machine_id, hostname, display);
```

Comment:
Is there any reason to expect that there will be more than one file in `ibus/bus`? If not, you could avoid a great deal of effort here and just open the only file in that directory, if it exists.

Reply:
@atheriel
Actually I have more than one file in there since $DISPLAY isn't constant and ibus doesn't clear the folder at any stage.

Reply:
If it's just `$DISPLAY` we expect to change we can just look for the file matching that, right? No need to muck around by guessing the hostname and machine ID. Or do users end up with more than one machine ID?

We could also let (or expect that) users set a `socket_path` manually in their configuration.

Reply:
Actually that is a good point. So can simplify the code by looking at the directory conents and pattern match 32charstring-string-$DISPLAY. Or perhaps first check the directory and if there is only one file just go with that, otherwise if more than one file then pattern match to find the right one.

> The machine ID is usually generated from a random source during system installation or first boot and stays constant for all subsequent boots. Optionally, for stateless systems, it is generated during runtime during early boot if necessary. The machine ID is a single newline-terminated, hexadecimal, 32-character, lowercase ID.